libsupport/include/katana/OpaqueID.h

```diff
@@ -278,6 +302,83 @@ struct OpaqueIDLinear : public OpaqueIDOrderedWithValue<_IDType, _Value> {
   }
 };
 
+
+/// A Counting Iterator for OpaqueIDLinear
+template <typename IDType>
+class OpaqueIDCountingIterator: 
+  public boost::iterator_facade<
+    OpaqueIDCountingIterator<IDType>, // derived class
+    IDType, // Value type
+    std::random_access_iterator_tag> // iterator_category
+    
+{
+  friend class boost::iterator_core_access;
+
+  using Base = boost::iterator_facade<
+    OpaqueIDCountingIterator, IDType, std::random_access_iterator_tag>;
+
+  static_assert(std::is_base_of_v<
+        OpaqueIDLinear<IDType, typename IDType::underlying_type>, 
+        IDType
+        >,
+      "OpaqueIDCountingIterator can only be created for OpaqueIDLinear");
+
+
+public:
+
+  OpaqueIDCountingIterator() noexcept = default;
+  explicit OpaqueIDCountingIterator(const IDType& id) noexcept: value_(id) {}
+
+  OpaqueIDCountingIterator(const OpaqueIDCountingIterator& other) noexcept = default;
+  OpaqueIDCountingIterator(OpaqueIDCountingIterator&& other) noexcept = default;
+
+  ~OpaqueIDCountingIterator() noexcept = default;
+
+  OpaqueIDCountingIterator& operator=(const OpaqueIDCountingIterator& other) noexcept = default;
+  OpaqueIDCountingIterator& operator=(OpaqueIDCountingIterator&& other) noexcept = default;
+
+private:
+
+  bool equal(const OpaqueIDCountingIterator& that) const noexcept {
+    return value_ == that.value_;
+  }
+
+  typename Base::reference
```

Comment:
Do you need to add `const` here?

Reply:
There's a boost library issue here that prevents me from returning const ref. It's the way the iterator_facade code is written.